core/src/main/java/org/elasticsearch/index/percolator/PercolatorFieldMapper.java

```diff
@@ -60,19 +73,28 @@ public Builder(QueryShardContext queryShardContext) {
         @Override
         public PercolatorFieldMapper build(BuilderContext context) {
             context.path().add(name);
-            KeywordFieldMapper extractedTermsField = createStringFieldBuilder(EXTRACTED_TERMS_FIELD_NAME).build(context);
-            KeywordFieldMapper unknownQueryField = createStringFieldBuilder(UNKNOWN_QUERY_FIELD_NAME).build(context);
+            KeywordFieldMapper extractedTermsField = createExtractQueryFieldBuilder(EXTRACTED_TERMS_FIELD_NAME).build(context);
+            KeywordFieldMapper unknownQueryField = createExtractQueryFieldBuilder(UNKNOWN_QUERY_FIELD_NAME).build(context);
+            KeywordFieldMapper queryBuilderField = createQueryBuilderFieldBuilder().build(context);
             context.path().remove();
-            return new PercolatorFieldMapper(name(), fieldType, defaultFieldType, context.indexSettings(), multiFieldsBuilder.build(this, context), copyTo, queryShardContext, extractedTermsField, unknownQueryField);
+            return new PercolatorFieldMapper(name(), fieldType, defaultFieldType, context.indexSettings(), multiFieldsBuilder.build(this, context), copyTo, queryShardContext, extractedTermsField, unknownQueryField, queryBuilderField);
         }
 
-        static KeywordFieldMapper.Builder createStringFieldBuilder(String name) {
+        static KeywordFieldMapper.Builder createExtractQueryFieldBuilder(String name) {
             KeywordFieldMapper.Builder queryMetaDataFieldBuilder = new KeywordFieldMapper.Builder(name);
             queryMetaDataFieldBuilder.docValues(false);
             queryMetaDataFieldBuilder.store(false);
             queryMetaDataFieldBuilder.indexOptions(IndexOptions.DOCS);
             return queryMetaDataFieldBuilder;
         }
+
+        static KeywordFieldMapper.Builder createQueryBuilderFieldBuilder() {
+            KeywordFieldMapper.Builder queryMetaDataFieldBuilder = new KeywordFieldMapper.Builder(QUERY_BUILDER_FIELD_NAME);
+            queryMetaDataFieldBuilder.docValues(false);
+            queryMetaDataFieldBuilder.store(true);
```

Comment:
I also like that it makes the percolator more predictable. Thinking about it more... I realized that my suggestion to keep parsing the source is also hard because it would require to keep track of paths since percolator queries can be anywhere in the source now. So I think I would agree that having a dedicated field for loading makes things simpler. But then let's use a binary doc values field instead of a stored field, this will make the loading more efficient, and if we ever decide to keep queries on disk, the data structure will already be there?